From: Goldwyn Rodrigues <rgoldwyn@suse.de>
Subject: Fix oops in set_map_drive
References: bnc#446824, bnc#444469

The oops was caused because of an unconditional free because of the
merge changes.
The error was caused because novfs_set_map_drive was not being called
with the right args, which caused it to request for incorrect memory
size.
Cleaned up some debug messages as well, and corrected debug messages.


---
 fs/novfs/daemon.c |    7 ++++---
 fs/novfs/inode.c  |   11 -----------
 2 files changed, 4 insertions(+), 14 deletions(-)

--- a/fs/novfs/daemon.c
+++ b/fs/novfs/daemon.c
@@ -1894,7 +1894,7 @@ static int set_map_drive(struct novfs_xp
 		full_name_hash(drivemap->name,
 				symInfo.linkOffsetLength - 1);
 	drivemap->namelen = symInfo.linkOffsetLength - 1;
-	DbgPrint("hash=0x%x path=%s", drivemap->hash, drivemap->name);
+	DbgPrint("hash=0x%lx path=%s", drivemap->hash, drivemap->name);
 
 	dm = (struct drive_map *) & DriveMapList.next;
 
@@ -1903,7 +1903,7 @@ static int set_map_drive(struct novfs_xp
 	list_for_each(list, &DriveMapList) {
 		dm = list_entry(list, struct drive_map, list);
 		__DbgPrint("%s: dm=0x%p\n"
-				"   hash:    0x%x\n"
+				"   hash:    0x%lx\n"
 				"   namelen: %d\n"
 				"   name:    %s\n", __func__,
 				dm, dm->hash, dm->namelen, dm->name);
@@ -1928,7 +1928,8 @@ static int set_map_drive(struct novfs_xp
 					&dm->list);
 		}
 	}
-	kfree(drivemap);
+	else
+		kfree(drivemap);
 	up(&DriveMapLock);
 	return (retVal);
 }
--- a/fs/novfs/inode.c
+++ b/fs/novfs/inode.c
@@ -4010,22 +4010,11 @@ int __init init_novfs(void)
 
 void __exit exit_novfs(void)
 {
-	printk(KERN_INFO "exit_novfs\n");
-
 	novfs_scope_exit();
-	printk(KERN_INFO "exit_novfs after Scope_Uninit\n");
-
 	novfs_daemon_queue_exit();
-	printk(KERN_INFO "exit_novfs after Uninit_Daemon_Queue\n");
-
 	novfs_profile_exit();
-	printk(KERN_INFO "exit_novfs after profile_exit\n");
-
 	novfs_proc_exit();
-	printk(KERN_INFO "exit_novfs Uninit_Procfs_Interface\n");
-
 	unregister_filesystem(&novfs_fs_type);
-	printk(KERN_INFO "exit_novfs: Exit\n");
 
 	if (novfs_current_mnt) {
 		kfree(novfs_current_mnt);
